src/sql/src/query_model/lowering.rs

```diff
@@ -143,7 +137,71 @@ impl<'a> Lowerer<'a> {
                         .collect_vec(),
                 )
             }
+            BoxType::Grouping(grouping) => {
+                // Note: a grouping box must only contain a single quantifier but we can still
+                // re-use `lower_join` for single quantifier joins
+                let (mut input, column_map) =
+                    self.lower_join(get_outer, outer_column_map, &the_box.quantifiers, id_gen);
+
+                // Build the reduction
+                let group_key = grouping
+                    .key
+                    .iter()
+                    .map(|k| Self::lower_expression(k, &column_map))
+                    .collect_vec();
+                let aggregates = the_box
+                    .columns
+                    .iter()
+                    .filter_map(|c| {
+                        if let BoxScalarExpr::Aggregate {
+                            func,
+                            expr,
+                            distinct,
+                        } = &c.expr
+                        {
+                            Some(expr::AggregateExpr {
+                                func: func.clone(),
+                                expr: Self::lower_expression(expr, &column_map),
+                                distinct: *distinct,
+                            })
+                        } else {
+                            None
+                        }
+                    })
+                    .collect_vec();
+                input = SR::Reduce {
+                    input: Box::new(input),
+                    group_key,
+                    aggregates,
+                    monotonic: false,
+                    expected_group_size: None,
+                };
+
+                // Put the columns in the same order as projected by the Grouping box by
+                // adding an additional projection
+                let mut aggregate_count = 0;
+                let projection = the_box.columns.iter().map(|c| {
```

Comment:
When does the grouping box not project columns in the order `<group_keys>, <aggregates>`? Does this only happen once QGM parses the SQL directly as opposed to takes in an HIR?

Reply:
Currently it is not possible. But it could be, for example, the result of a transformation. We could realize that:

```
select a, max(b), max(a) from t group by a;
```

is equivalent to

```
select a, max(b), a from t group by a;
```

and end up with a `Grouping` box with that projection after simplifying it.

Reply:
Makes sense. Can you just add a comment that this functionality is not strictly necessary right now and when it may be needed?

Reply:
Sure. Done